homeassistant/components/bring/todo.py

```diff
@@ -0,0 +1,186 @@
+from typing import TYPE_CHECKING
+
+from python_bring_api.exceptions import BringRequestException
+from python_bring_api.types import BringList
+
+from homeassistant import config_entries, core
+from homeassistant.components.todo import (
+    TodoItem,
+    TodoItemStatus,
+    TodoListEntity,
+    TodoListEntityFeature,
+)
+from homeassistant.core import callback
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from .const import DOMAIN
+from .coordinator import BringDataUpdateCoordinator
+
+_LOGGER = logging.getLogger(__name__)
+
+SCAN_INTERVAL = timedelta(seconds=90)
+
+
+async def async_setup_entry(
+    hass: core.HomeAssistant,
+    config_entry: config_entries.ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the sensor from a config entry created in the integrations UI."""
+    coordinator = hass.data[DOMAIN][config_entry.entry_id]
+
+    unique_id = config_entry.unique_id
+
+    if TYPE_CHECKING:
+        assert unique_id
+
+    for _list in coordinator.data:
+        async_add_entities(
+            [
+                BringTodoListEntity(
+                    coordinator,
+                    name=f"Bring! {_list['name']}",
+                    unique_id=f"{unique_id} {_list['listUuid']}",
+                )
+            ],
+        )
+
+
+class BringTodoListEntity(
+    CoordinatorEntity[BringDataUpdateCoordinator], TodoListEntity
+):
+    """A To-do List representation of the Bring! Shopping List."""
+
+    _attr_has_entity_name = True
+    _attr_supported_features = (
+        TodoListEntityFeature.CREATE_TODO_ITEM
+        | TodoListEntityFeature.UPDATE_TODO_ITEM
+        | TodoListEntityFeature.DELETE_TODO_ITEM
+        | TodoListEntityFeature.SET_DESCRIPTION_ON_ITEM
+    )
+    _attr_should_poll = False
+
+    def __init__(
+        self,
+        coordinator: BringDataUpdateCoordinator,
+        name: str,
+        unique_id: str,
+    ) -> None:
+        """Initialize BringTodoListEntity."""
+        super().__init__(coordinator)
+        self._attr_icon = "mdi:cart"
+        self._attr_name = name
+        self._attr_unique_id = unique_id
+
+    async def async_added_to_hass(self) -> None:
+        """Populate the todo list at start."""
+        self._populate_list()
+        await super().async_added_to_hass()
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle the state update and prepare the todo list items."""
+        self._populate_list()
+        return super()._handle_coordinator_update()
+
+    @property
+    def list_(self) -> BringList | None:
+        """Return the bring list."""
+        return next(
+            (
+                _list
+                for _list in self.coordinator.data
+                if _list["listUuid"] == self.unique_id
+            ),
+            None,
+        )
```

Comment:
Why do we do this?

Reply:
To get the correct shopping list from the coordinator, which loads the data for all bring lists at once.